### ISSUE TITLE
`or_fun_call` false positive on nightly with some enums

### ISSUE THREAD

**Issue by cottnn** [rust-lang/rust-clippy#9693]
### Summary

The `or_fun_call` lint can be triggered even if the argument to `.or` is a regular enum variant. From my tests, this only happens when the enum has additional variants with certain value types, such as `io::Error`. I tested some other value types, and it looks like it depends on whether the type implements `Clone` or not. So putting `Vec<u8>` in one of the enum variants will not trigger the lint, but putting `io::Error` (which doesn't implement `Clone`) will trigger the lint. I could be completely wrong on this one though

### Lint Name

or_fun_call

### Reproducer

I tried this code:

```rust
use std::io;

enum MyError {
    A,
    B(io::Error)
}

fn main() {
    Ok::<(), MyError>(()).or(Err(MyError::A));
}
```

I saw this happen:

```
warning: use of `or` followed by a function call
 --> src/main.rs:9:27
  |
9 |     Ok::<(), MyError>(()).or(Err(MyError::A));
  |                           ^^^^^^^^^^^^^^^^^^^ help: try this: `or_else(|_| Err(MyError::A))`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
  = note: `#[warn(clippy::or_fun_call)]` on by default
```

I expected to see this happen:
No clippy lint

### Version

```text
rustc 1.66.0-nightly (5c8bff74b 2022-10-21)
binary: rustc
commit-hash: 5c8bff74bc1c52bef0c79f3689bb227f51f3e82d
commit-date: 2022-10-21
host: x86_64-unknown-linux-gnu
release: 1.66.0-nightly
LLVM version: 15.0.2
```

**Comment by kraktus** [comment]
Should be fixed on next nightly when https://github.com/rust-lang/rust-clippy/pull/9610 lands

**Comment by J-ZhengLi** [comment]
> Should be fixed on next nightly when #9610 lands

Looks like this was indeed fixed 👀. closing now~
